### DISCUSSION TITLE
contrib repository for strict http middleware

### DISCUSSION THREAD

**Discussion by emilien-puget** [oapi-codegen/oapi-codegen#1655] [Ideas]
it would be nice to have a contrib repository to host strict http middleware

**Comment by jamietanna** [comment]
Interesting @emilien-puget - what sort of middleware are you thinking here? As in the same middleware for http://github.com/oapi-codegen/nethttp-middleware but with strict bindings? Or common middleware that could be used with the strict server?

**Reply by emilien-puget** [reply]
I am thinking middleware like an access logger, open telemetry.

Using the strict server middleware gives us more information, mainly the operation Id

**Reply by felix** [reply]
If this helps, I knocked up something like this:
```go
func otelMiddleware(next apiDef.StrictHandlerFunc, operationID string) apiDef.StrictHandlerFunc {
        props := otel.GetTextMapPropagator()
        return func(ctx context.Context, w http.ResponseWriter, r *http.Request, request any) (response any, err error) {
                rctx := props.Extract(r.Context(), propagation.HeaderCarrier(r.Header))
                var tracer trace.Tracer
                if span := trace.SpanFromContext(r.Context()); span.SpanContext().IsValid() {
                        tracer = span.TracerProvider().Tracer(tracerName)
                } else {
                        tracer = otel.GetTracerProvider().Tracer(tracerName)
                }
                _, span := tracer.Start(rctx, operationID, trace.WithSpanKind(trace.SpanKindServer))
                defer span.End()
                resp, err := next(trace.ContextWithSpan(ctx, span), w, r, request)
                return resp, err
        }
}
```
but I should probably expand it to include timings etc.

**Comment by mromaszewicz** [comment]
The middleware works at a more generic level, and I don't think we need a different one for strict clients versus the more flexible ones, as the underlying router is exactly the same. Middleware is router specific, so we need it to be special cased for echo, gin, etc, but not for the client type.

I've also intentionally tried to keep the scope of this project minimal. I'm most familiar with echo and gin myself, and we can just use their off-the-shelf middleware without re-implementing ourselves. If possible, I think you should look around if others have shared the sort of middleware that you are looking for. OperationID, for example, is totally 1:1 with the URI being called, and you can get that from the openapi3filter router.

**Comment by emilien-puget** [comment]
what prompted this suggestion is that we can't use some standard middleware as is, such as https://github.com/open-telemetry/opentelemetry-go-contrib/blob/main/instrumentation/net/http/otelhttp/example/server/server.go#L91 as it requires to wrap each handle separately to pass information such as the OperationId
